docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb

```diff
@@ -0,0 +1,1265 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {
+      "byteLimit": 2048000,
+      "rowLimit": 10000
+     },
+     "inputWidgets": {},
+     "nuid": "fa602d14-6372-4e0a-92d8-6f7db048c3b8",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "[DSPy](https://dspy-docs.vercel.app/) simplifies building language model (LM) pipelines by replacing manual prompt engineering with structured \"text transformation graphs.\" These graphs use flexible, learning modules that automate and optimize LM tasks like reasoning, retrieval, and answering complex questions. \n",
+    "\n",
+    "### How does it work?\n",
+    "At a high level, DSPy optimizes prompts, selects the best language model, and can even fine-tune the model using training data.\n",
+    "\n",
+    "The process follows these three steps, common to most DSPy [optimizers](https://dspy-docs.vercel.app/docs/building-blocks/optimizers):\n",
+    "\n",
+    "1. **Candidate Generation**: DSPy finds all `Predict` modules in the program and generates variations of instructions and demonstrations (e.g., examples for prompts). This step creates a set of possible candidates for the next stage.\n",
+    "2. **Parameter Optimization**: DSPy then uses methods like random search, TPE, or Optuna to select the best configurations. Fine-tuning models can also be done at this stage.\n",
+    "3. **Higher-Order Optimization**: Lastly, DSPy can modify the program's structure, such as creating ensembles of different variations of the program to boost performance.\n",
```

Comment:
hmm which optimizer does this?

Reply:
Great callout. According to [this paper](https://arxiv.org/pdf/2406.11695) temperature on the `proposer LM`, not the optimized LM. I misinterpreted and will fix.

```
0-Shot MIPRO++ tunes instructions by metaoptimizing how or whether our Grounded instruction proposal strategy uses the dataset summary
(boolean), uses the program summary (boolean),
adjusts the proposer LM temperature (float), provides the proposer LM a plain-text tip on prompt
engineering (categorical; see Appendix C), and
selects a specific set of bootstrapped demos to show
to the proposer LM (categorical). A Bayesian
model with the same mini-batching strategy employed in MIPRO is then used to optimize over
these hyperparameters. A program with the newly
proposed instruction is evaluated on each trial, using the mini-batching approach described earlier.
The best fully-evaluated program is returned.
```